File: bindings/java/src/main/java/io/oucs/OucsStream.java
package io.oucs;

/**
 * OucsStream — chunk-by-chunk low-memory audio streaming.
 *
 * <pre>{@code
 * try (OucsStream s = reader.stream(0, 4096)) {
 *     byte[] chunk;
 *     while ((chunk = s.readChunk()) != null) {
 *         audioPlayer.feed(chunk);
 *     }
 * }
 * }</pre>
 */
public class OucsStream implements AutoCloseable {

    private long _ptr;
    private final int _chunkSize;

    OucsStream(long readerPtr, int songIndex, int chunkSize, String password) {
        _chunkSize = chunkSize;
        _ptr = nativeOpen(readerPtr, songIndex, chunkSize, password);
        if (_ptr == 0) throw new OucsException("Cannot open stream for song " + songIndex);
    }

    /**
     * Read the next chunk. Returns null at end of stream.
     * Only this chunk is in memory — device load is near zero.
     */
    public byte[] readChunk() {
        if (_ptr == 0) return null;
        return nativeReadChunk(_ptr, _chunkSize);
    }

    /**
     * Seek to a byte offset within the song's audio data.
     */
    public void seek(long byteOffset) {
        OucsException.checkCode(nativeSeek(_ptr, byteOffset), "seek");
    }

    /**
     * Seek to a chapter by index.
     */
    public void seekChapter(int chapterIndex) {
        OucsException.checkCode(nativeSeekChapter(_ptr, chapterIndex), "seekChapter");
    }

    /**
     * Current byte position within the song.
     */
    public long getPosition() {
        return _ptr != 0 ? nativeTell(_ptr) : 0;
    }

    /**
     * Read all audio data into a single byte array.
     * Use with care for large files.
     */
    public byte[] readAll() {
        java.io.ByteArrayOutputStream baos = new java.io.ByteArrayOutputStream();
        byte[] chunk;
        while ((chunk = readChunk()) != null) {
            baos.write(chunk, 0, chunk.length);
        }
        return baos.toByteArray();
    }

    @Override
    public void close() {
        if (_ptr != 0) { nativeFree(_ptr); _ptr = 0; }
    }

    // ── Native declarations ───────────────────────────────────
    private static native long   nativeOpen(long readerPtr, int songIndex,
                                             int chunkSize, String password);
    private static native byte[] nativeReadChunk(long ptr, int chunkSize);
    private static native int    nativeSeek(long ptr, long offset);
    private static native int    nativeSeekChapter(long ptr, int chapterIndex);
    private static native long   nativeTell(long ptr);
    private static native void   nativeFree(long ptr);
}

File: bindings/javascript/src/OucsStream.js
/**
 * OucsStream — async chunk-by-chunk song streaming.
 * Implements AsyncIterable so you can use for-await-of.
 */
'use strict';

const { getModule }   = require('./wasm');
const { raiseForCode } = require('./errors');

const DEFAULT_CHUNK = 4096;

class OucsStream {
  /**
   * @param {number} readerPtr  - WASM pointer to OucsReader
   * @param {number} songIndex  - Song index to stream
   * @param {number} chunkSize  - Bytes per chunk
   * @param {string|null} password - Decryption password
   */
  constructor(readerPtr, songIndex, chunkSize = DEFAULT_CHUNK, password = null) {
    this._readerPtr = readerPtr;
    this._songIndex = songIndex;
    this._chunkSize = chunkSize;
    this._password  = password;
    this._ptr       = null;
    this._bufPtr    = null;
    this._m         = null;
  }

  async _init() {
    this._m = getModule();
    const passPtr = this._password
      ? this._m.allocateUTF8(this._password) : 0;

    this._ptr = this._m._oucs_stream_open(
      this._readerPtr, this._songIndex, this._chunkSize, passPtr
    );
    if (passPtr) this._m._free(passPtr);
    if (!this._ptr) throw new Error(`Cannot open stream for song ${this._songIndex}`);

    this._bufPtr = this._m._malloc(this._chunkSize);
    if (!this._bufPtr) throw new Error('Out of WASM memory');
    return this;
  }

  /** Read next chunk. Returns Uint8Array or null at EOF. */
  readChunk() {
    if (!this._ptr) return null;
    const sizePtr = this._m._malloc(4);
    const ret = this._m._oucs_stream_read_chunk(
      this._ptr, this._bufPtr, this._chunkSize, sizePtr
    );
    raiseForCode(ret, 'stream_read_chunk');
    const bytesRead = this._m.HEAP32[sizePtr >> 2];
    this._m._free(sizePtr);
    if (bytesRead === 0) return null;
    // Copy from WASM heap
    return new Uint8Array(this._m.HEAPU8.buffer, this._bufPtr, bytesRead).slice();
  }

  /** Seek to byte offset within song. */
  seek(byteOffset) {
    raiseForCode(this._m._oucs_stream_seek(this._ptr, byteOffset), 'stream_seek');
  }

  /** Seek to chapter by index. */
  seekChapter(chapterIndex) {
    raiseForCode(
      this._m._oucs_stream_seek_chapter(this._ptr, chapterIndex),
      'stream_seek_chapter'
    );
  }

  /** Current byte position. */
  get position() {
    return this._ptr ? this._m._oucs_stream_tell(this._ptr) : 0;
  }

  close() {
    if (this._ptr)    { this._m._oucs_stream_free(this._ptr); this._ptr = null; }
    if (this._bufPtr) { this._m._free(this._bufPtr); this._bufPtr = null; }
  }

  /** AsyncIterator — yields Uint8Array chunks. */
  async *[Symbol.asyncIterator]() {
    try {
      while (true) {
        const chunk = this.readChunk();
        if (!chunk) break;
        yield chunk;
      }
    } finally {
      this.close();
    }
  }

  /** Read all audio data into a single Uint8Array. */
  async readAll() {
    const chunks = [];
    let total = 0;
    for await (const chunk of this) {
      chunks.push(chunk);
      total += chunk.length;
    }
    const result = new Uint8Array(total);
    let offset = 0;
    for (const chunk of chunks) {
      result.set(chunk, offset);
      offset += chunk.length;
    }
    return result;
  }
}

module.exports = { OucsStream };
